### PR TITLE
chore: deprecate pytorch.reset_parameters [DET-3578]

### PR DESCRIPTION
**Pull request by rb-determined-ai**
# Description

We do not need to have users call `reset_parameters`, except for very specialized models (where the user needs to call `reset_parameters` on their own regardless of our system).  It should therefore not be a part of our API, since it only clutters our API for 99% of users.

The reason we do not need to call it is that `reset_parameters()` (or some similarly-named equivalent... they're not all named the same) is called at the end of `__init__()` for every built-in `nn.Module` in PyTorch.

## Appearances of `def .*reset.*param` in PyTorch 1.5

Found via `git grep -El 'def .*reset.*param'`:

### `torch/nn/intrinsic/qat/modules/conv_fused.py`

Note that `_ConvNd.__init__()` is going to call the overridden
`_ConvBnNd.reset_parameters`, which is what the `hasttr(self, 'gamma')` hack is
going to protect against.

```python
class _ConvBnNd(nn.modules.conv._ConvNd):
    def __init__(...):
        nn.modules.conv._ConvNd.__init__(...)
        ...
        self.gamma = nn.Parameter(torch.Tensor(out_channels))
        ...
        self.reset_bn_parameters()

    def reset_running_stats(self):
        self.running_mean.zero_()
        self.running_var.fill_(1)
        self.num_batches_tracked.zero_()

    def reset_bn_parameters(self):
        self.reset_running_stats()
        init.uniform_(self.gamma)
        init.zeros_(self.beta)

    def reset_parameters(self):
        super(_ConvBnNd, self).reset_parameters()
        # A hack to avoid resetting on undefined parameters
        if hasattr(self, 'gamma'):
            self.reset_bn_parameters()
```

### `torch/nn/modules/activation.py`

Notice that the `reset_parameters()` is actually called `_reset_parameters()`.

```python
class MultiheadAttention(Module):
    def __init__(...):
        super(MultiheadAttention, self).__init__()
        ...
        self._reset_parameters()

    def _reset_parameters(self):
        ...
```

### `torch/nn/modules/adaptive.py`

Notice that the `head` and each `tail` are all `Linear`s which get initialized
automatically, but `reset_parameters` is still definied if you want to
reinitialize them.

```python
class AdaptiveLogSoftmaxWithLoss(Module):
    def __init__(...):
        super(AdaptiveLogSoftmaxWithLoss, self).__init__()
        ...
        self.head = Linear(self.in_features, self.head_size, bias=self.head_bias)
        self.tail = ModuleList()

        for i in range(self.n_clusters):

            hsz = int(self.in_features // (self.div_value ** (i + 1)))
            osz = self.cutoffs[i + 1] - self.cutoffs[i]

            projection = Sequential(
                Linear(self.in_features, hsz, bias=False),
                Linear(hsz, osz, bias=False)
            )

            self.tail.append(projection)

    def reset_parameters(self):
        self.head.reset_parameters()
        for i2h, h2o in self.tail:
            i2h.reset_parameters()
            h2o.reset_parameters()
```

### `torch/nn/modules/batchnorm.py`

```python
class _NormBase(Module):
    def __init__(...):
        super(_NormBase, self).__init__()
        ...
        self.reset_parameters()

    def reset_running_stats(self):
        if self.track_running_stats:
            self.running_mean.zero_()
            self.running_var.fill_(1)
            self.num_batches_tracked.zero_()

    def reset_parameters(self):
        self.reset_running_stats()
        if self.affine:
            init.ones_(self.weight)
            init.zeros_(self.bias)
```

### `torch/nn/modules/conv.py`

```python
class _ConvNd(Module):
    def __init__(...):
        super(_ConvNd, self).__init__()
        ...
        self.reset_parameters()

    def reset_parameters(self):
        ...
```

### `torch/nn/modules/linear.py`

```python
class Linear(Module):
    def __init__(...):
        super(Linear, self).__init__()
        ...
        self.reset_parameters()

    def reset_parameters(self):
        init.kaiming_uniform_(self.weight, a=math.sqrt(5))
        if self.bias is not None:
            fan_in, _ = init._calculate_fan_in_and_fan_out(self.weight)
            bound = 1 / math.sqrt(fan_in)
            init.uniform_(self.bias, -bound, bound)

...

class Bilinear(Module):
    def __init__(...):
        super(Bilinear, self).__init__()
        ...
        self.reset_parameters()

    def reset_parameters(self):
        bound = 1 / math.sqrt(self.weight.size(1))
        init.uniform_(self.weight, -bound, bound)
        if self.bias is not None:
            init.uniform_(self.bias, -bound, bound)
```

### `torch/nn/modules/normalization.py`

```python
class LayerNorm(Module):
    def __init__(...):
        super(LayerNorm, self).__init__()
        ...
        self.reset_parameters()

    def reset_parameters(self):
        if self.elementwise_affine:
            init.ones_(self.weight)
            init.zeros_(self.bias)

...

class GroupNorm(Module):
    def __init__(...):
        super(GroupNorm, self).__init__()
        ...
        self.reset_parameters()

    def reset_parameters(self):
        if self.affine:
            init.ones_(self.weight)
            init.zeros_(self.bias)
```

### `torch/nn/modules/rnn.py`

```python
class RNNBase(Module):
    def __init__(...):
        super(RNNBase, self).__init__()
        ...
        self.reset_parameters()

    def __setattr__(self, attr, value):
        if hasattr(self, "_flat_weights_names") and attr in self._flat_weights_names:
            # keep self._flat_weights up to date if you do self.weight = ...
            idx = self._flat_weights_names.index(attr)
            self._flat_weights[idx] = value
        super(RNNBase, self).__setattr__(attr, value)

    def reset_parameters(self):
        stdv = 1.0 / math.sqrt(self.hidden_size)
        for weight in self.parameters():
            init.uniform_(weight, -stdv, stdv)

...


class RNNCellBase(Module):
    def __init__(...):
        super(RNNCellBase, self).__init__()
        self.reset_parameters()

    def reset_parameters(self):
        stdv = 1.0 / math.sqrt(self.hidden_size)
        for weight in self.parameters():
            init.uniform_(weight, -stdv, stdv)
```

### `torch/nn/modules/sparse.py`

```python
class Embedding(Module):
    def __init__(self, num_embeddings, embedding_dim, padding_idx=None,
                 max_norm=None, norm_type=2., scale_grad_by_freq=False,
                 sparse=False, _weight=None):
        super(Embedding, self).__init__()
        ...
        if _weight is None:
            self.weight = Parameter(torch.Tensor(num_embeddings, embedding_dim))
            self.reset_parameters()
        else:
            assert list(_weight.shape) == [num_embeddings, embedding_dim], \
                'Shape of weight does not match num_embeddings and embedding_dim'
            self.weight = Parameter(_weight)
        ...

    def reset_parameters(self):
        init.normal_(self.weight)
        if self.padding_idx is not None:
            with torch.no_grad():
                self.weight[self.padding_idx].fill_(0)

...

class EmbeddingBag(Module):
    def __init__(self, num_embeddings, embedding_dim,
                 max_norm=None, norm_type=2., scale_grad_by_freq=False,
                 mode='mean', sparse=False, _weight=None, include_last_offset=False):
        super(EmbeddingBag, self).__init__()
        ...
        if _weight is None:
            self.weight = Parameter(torch.Tensor(num_embeddings, embedding_dim))
            self.reset_parameters()
        else:
            assert list(_weight.shape) == [num_embeddings, embedding_dim], \
                'Shape of weight does not match num_embeddings and embedding_dim'
            self.weight = Parameter(_weight)
        ...

    def reset_parameters(self):
        init.normal_(self.weight)
```

### `torch/nn/modules/transformer.py`

Notice that the `reset_parameters()` is actually called `_reset_parameters()`.

```python
class Transformer(Module):
    def __init__(...):
        super(Transformer, self).__init__()
        ...
        self._reset_parameters()
        ...

    def _reset_parameters(self):
        for p in self.parameters():
            if p.dim() > 1:
                xavier_uniform_(p)
```

### CODE REVIEW

**Comment by rb-determined-ai** [comment]
Update: I failed the XOR unit tests due to affecting the random seeds differently.  We talked about replacing the nonlinear unit tests with linear tests a long time ago (last october), so I took the liberty of replacing the failing tests with a linear model test.  I consider that work to be a step in the direction of completing https://determinedai.atlassian.net/browse/DET-1597.

**Comment by rb-determined-ai** [comment]
@shiyuann please review the unit test update, specifically.

File: harness/tests/experiment/fixtures/pytorch_onevar_model.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        assert float_eq(
+            metrics["loss"], metrics["loss_exp"]
+        ), f'{metrics["loss"]} does not match {metrics["loss_exp"]} at batch {batch_idx}'
+
+        assert float_eq(
+            metrics["w_after"], metrics["w_exp"]
+        ), f'{metrics["w_after"]} does not match {metrics["w_exp"]} at batch {batch_idx}'
+
+    def evaluate_batch(self, batch: pytorch.TorchData) -> Dict[str, Any]:
+        # Return the loss.
```
Comment (shiyuann) [inline harness/tests/experiment/fixtures/pytorch_onevar_model.py:115]: non-blocking: unnecessary comment.

File: harness/tests/experiment/fixtures/pytorch_onevar_model.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+expecting.
+
+let:
+    l = loss
+    w = the value of the one weight
+    R = learning rate (constant)
+
+then calculate the loss:
+
+(1)     l = (label - (data * w)) ** 2
```
Comment (shiyuann) [inline harness/tests/experiment/fixtures/pytorch_onevar_model.py:18]: non-blocking: Better put down the initial values `w0`.

Comment (rb-determined-ai) [reply harness/tests/experiment/fixtures/pytorch_onevar_model.py:18]: done

File: harness/tests/experiment/fixtures/pytorch_onevar_model.py (hunk 1 above)
Comment (rb-determined-ai) [reply harness/tests/experiment/fixtures/pytorch_onevar_model.py:115]: removed.

File: harness/determined/pytorch/_pytorch_trial.py [hunk 3]
```diff
@@ -959,12 +959,20 @@ def evaluate_full_dataset(self, data_loader: torch.utils.data.DataLoader) -> Dic
...
+        nn.Modules which do need a call to reset_parameters(), it is recommended to call
+        self.reset_parameters() directly in their __init__() function, as is standard in all
+        built-in nn.Modules.
 
-    Important: Call this prior to loading any backbone weights,
-    otherwise those weights will be overwritten.
+    Recursively calls ``reset_parameters()`` for all modules.
     """
-    logging.info("Resetting model parameters.")
+    logging.warn(
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/pytorch/_pytorch_trial.py:970]: fixed
